crates/cubecl-core/src/frontend/branch.rs

```diff
@@ -408,6 +411,110 @@ pub fn if_else_expr_expand<C: CubePrimitive>(
     }
 }
 
+pub struct SwitchExpand<I: Int> {
+    value: ExpandElementTyped<I>,
+    default: CubeContext,
+    cases: Vec<(ExpandElementTyped<I>, CubeContext)>,
+}
+
+impl<I: Int> SwitchExpand<I> {
+    pub fn case(
+        mut self,
+        context: &mut CubeContext,
+        value: impl Int,
+        block: impl FnOnce(&mut CubeContext),
+    ) -> Self {
+        let value = I::from(value).unwrap();
+        let mut case_child = context.child();
+        block(&mut case_child);
+        self.cases.push((value.into(), case_child));
+        self
+    }
+
+    pub fn finish(self, context: &mut CubeContext) {
+        let value_var = *self.value.expand;
+        context.register(Branch::Switch(Switch {
+            value: value_var,
+            scope_default: self.default.into_scope(),
+            cases: self
+                .cases
+                .into_iter()
+                .map(|it| (*it.0.expand, it.1.into_scope()))
+                .collect(),
+        }));
+    }
+}
+
+pub fn switch_expand<I: Int>(
+    context: &mut CubeContext,
+    value: ExpandElementTyped<I>,
+    default_block: impl FnOnce(&mut CubeContext),
+) -> SwitchExpand<I> {
+    let mut default_child = context.child();
+    default_block(&mut default_child);
+
+    SwitchExpand {
+        value,
+        default: default_child,
+        cases: Vec::new(),
+    }
+}
+
+pub struct SwitchExpandExpr<I: Int, C: CubePrimitive> {
+    value: ExpandElementTyped<I>,
+    out: ExpandElementTyped<C>,
+    default: CubeContext,
+    cases: Vec<(ExpandElementTyped<I>, CubeContext)>,
+}
```

Comment:
`SwitchExpandExpr` creates an outer output variables and adds an assign statement to each branch to assign the return to the outer variable. It's the same as the `IfExpandExpr`, it's used to emulate Rust behaviour that allows using match as  value in assignments. For typing reasons and to avoid a pointless assign in every branch, the two are split based on whether the match has a return value.